### ISSUE TITLE
While Dragging, "pointer-events" on children should be "none"

### ISSUE THREAD

**Issue by bradennapier** [palantir/react-mosaic#79]
In my case the widgets themselves register various pointer-events (TradingView chart).  This makes dragging impossible since the second I end up over the chart they take over the pointer events completely and the drag/drop is immediately lost.

So when a drag/drop is started, we need to tell `react-mosaic-component` that all pointer-events of the panels should be ignored.

Either that or an event is required to allow us to react to `onDragStart` and `onDragEnd`

**Comment by bradennapier** [comment]
LoL just a bump here as I just spent a hour or two re-figuring this bug out only to find my own post detailing why it was happening :-P

**Comment by nomcopter** [comment]
Heh, that's funny. This would make the most sense to me as a props (maybe context) on `MosaicDragWindow`. I don't use this project personally anymore so updates coming from me will take a while, but I'm definitely accepting PRs!

**Comment by bradennapier** [comment]
Ah, interesting.  Well in that case - do you perhaps have a second to go over some of the overall things, based on what you've learned, that you might do differently or add if you were to start over with this project? 

I was close to doing a rewrite a few times now.  If you aren't actively maintaining (as quickly) I may look at doing a new version and/or updating the current depending on the needed changes/scope.

I would imagine there may be interesting methods of doing things with the new context api and friends that could be investigated here.

**Comment by nomcopter** [comment]
Overall I'm quite happy with its current state. Fixing #31 and #50 would require a fair amount of rethought around the overall tree structure, which might make sense to spend more time thinking about if starting anew - both involve more complexity though.

The new context API is fairly similar to the old one (and it would be straightforward to add the new one to `react-mosaic` [also accepting PRs ;)])

**Comment by bradennapier** [comment]
yeah #31 is one I also have had comments on from users.  What would you imagine the fix looks like there?  I played with that to try to build a PR but did realize how complex it would be :-P

**Comment by nomcopter** [comment]
I would change `MosaicParent` to look more like the following:
```ts
export interface MosaicParent<T extends MosaicKey> {
  direction: MosaicDirection;
  children: MosaicNode<T>[];
  splitPercentage?: number;
}
```
instead of the explicit `first`/`second`

**Comment by bradennapier** [comment]
Unfortunately the problem seems to be a lot more difficult than meets the eye.

![image](https://user-images.githubusercontent.com/15365418/50514397-bdd59d00-0a52-11e9-9132-879ee023a6fd.png)

 After some research into a similar data structure - I came up with this little format - blue lines are `nodes` and red are `grids`:

```javascript
{
        direction: "h",
        children: [
          {
            componentID: "a",
            weight: 40
          },
          {
            direction: "v",
            weight: 40,
            children: [
              {
                componentID: "b",
                weight: 20
              },
              {
                weight: 80,
                children: [
                  {
                    direction: "v",
                    weight: 50,
                    children: [
                      {
                        direction: "h",
                        weight: 50,
                        children: [
                          {
                            componentID: "c",
                            weight: 50
                          },
                          {
                            componentID: "ok",
                            weight: 50
                          }
                        ]
                      },
                      {
                        componentID: "hi",
                        weight: 50
                      }
                    ]
                  },
                  {
                    componentID: "d",
                    weight: 50
                  }
                ]
              }
            ]
          },
          {
            componentID: "e",
            weight: 20
          }
        ]
      }
```

Now the problem is that say you want to resize `E` from its left side.  We will expect that resizing `E` to be wider will only affect the width of `B` and `D`.  However, since `D` is nested in a grid and `B` is a parent of that - it would require calculating static widths here and manually adjusting `B` and `D` to move together.  

Just writing out loud here - its like 3 AM so my brain is mush but just seeing how it might need to look from a simpler standpoint for now.

**Comment by bradennapier** [comment]
@nomcopter i decided to have some fun and play with the various ideas given.  I did it from scratch as I'm not 100% versed in Typescript at this point (Flow user here) but I think the general effort could be helpful here.  

I haven't decided if I'm going to go through the pain of learning `React DnD` yet to implement the final piece but I do have some wants there as well in terms of how it will handle window placement (want to basically do it like VSCode does).

Provided the pieces it implements thus far.  I believe drag and drop is only missing piece.  It provides 60fps for most part during any resize as well.  A few optimizations still can be made but overall its pretty clean IMO! I used a few pieces from this as a reference as well (will obviously credit if i do decide to actually make it a lib).

Anyway, thought I'd share!  Would be interested in your thoughts on how the various requests were implemented though!

https://github.com/bradennapier/react-grid-poc
